utils/src/main/java/io/atomix/utils/concurrent/SingleThreadContext.java

```diff
@@ -29,82 +29,111 @@
 import java.util.concurrent.ThreadFactory;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
-
-import static com.google.common.base.Preconditions.checkState;
-import static io.atomix.utils.concurrent.Threads.namedThreads;
+import java.util.function.Consumer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * Single threaded context.
- * <p>
- * This is a basic {@link ThreadContext} implementation that uses a
- * {@link ScheduledExecutorService} to schedule events on the context thread.
+ *
+ * <p>This is a basic {@link ThreadContext} implementation that uses a {@link
+ * ScheduledExecutorService} to schedule events on the context thread.
  *
  * @author <a href="http://github.com/kuujo">Jordan Halterman</a>
  */
 public class SingleThreadContext extends AbstractThreadContext {
   protected static final Logger LOGGER = LoggerFactory.getLogger(SingleThreadContext.class);
-  private final ScheduledExecutorService executor;
-  private final Executor wrappedExecutor = new Executor() {
-    @Override
-    public void execute(Runnable command) {
-      try {
-        executor.execute(() -> {
+  private static final Consumer<Throwable> DEFAULT_UNCAUGHT_EXCEPTION_OBSERVER =
+      e -> LOGGER.error("An uncaught exception occurred", e);
+  protected final ScheduledExecutorService executor;
+  private final Consumer<Throwable> uncaughtExceptionObserver;
+  private final Executor wrappedExecutor =
+      new Executor() {
+        @Override
+        public void execute(Runnable command) {
           try {
-            command.run();
-          } catch (Exception e) {
-            LOGGER.error("An uncaught exception occurred", e);
+            executor.execute(() -> command.run());
+          } catch (RejectedExecutionException e) {
```

Comment:
Should we call the consumer here as well?

Reply:
I think it is safe to ignore it. This usually happens when the thread context is closed. It might be too much logs if we start logging it.